samples/nrf9160/modem_shell/src/drivers/nrf91_non_offloaded_dev.c

```diff
@@ -0,0 +1,507 @@
+	}
+}
+
+/******************************************************************************/
+
+static uint8_t *util_fake_dev_get_mac(struct nrf91_non_offload_dev_context *ctx)
+{
+	if (ctx->mac_addr[2] == 0x00) {
+		/* 00-00-5E-00-53-xx Documentation RFC 7042 */
+		ctx->mac_addr[0] = 0x00;
+		ctx->mac_addr[1] = 0x00;
+		ctx->mac_addr[2] = 0x5E;
+		ctx->mac_addr[3] = 0x00;
+		ctx->mac_addr[4] = 0x53;
+		ctx->mac_addr[5] = sys_rand32_get();
+	}
+
+	return ctx->mac_addr;
+}
+
+static void nrf91_non_offload_iface_init(struct net_if *iface)
+{
+	struct nrf91_non_offload_dev_context *ctx = net_if_get_device(iface)->data;
+	uint8_t *mac = util_fake_dev_get_mac(ctx);
+
+	ctx->iface = iface;
+	ctx->default_pdp_active = false;
+	ctx->mdm_skct_id = NO_MDM_SCKT;
+
+	/* The mac address is not really used but net if expects to find one. */
+	net_if_set_link_addr(iface, mac, 6, NET_LINK_ETHERNET);
+}
+
+static int nrf91_non_offload_init(const struct device *arg)
+{
+	int err = 0;
+
+	ARG_UNUSED(arg);
+
+	k_work_init_delayable(&mdm_socket_work, nrf91_non_offload_mdm_socket_worker);
+
+	err = link_shell_pdn_event_forward_cb_set(nrf91_non_offload_pdn_event_handler);
+	if (err) {
+		mosh_error("link_shell_pdn_event_forward_cb_set() failed, err %d\n", err);
+		return -1;
+	}
+	return 0;
+}
+
+/******************************************************************************/
+
+#define NRF91_MODEM_DATA_UL_BUFFER_SIZE 1500
+static char send_buffer[NRF91_MODEM_DATA_UL_BUFFER_SIZE];
+
+static int nrf91_non_offload_iface_send(const struct device *dev, struct net_pkt *pkt)
+{
+	int ret, data_len;
+
+	data_len = net_pkt_get_len(pkt);
+	ret = net_pkt_read(pkt, send_buffer, data_len);
+	if (ret < 0) {
+		mosh_error("%s: cannot read packet: %d, from pkt %p\n", __func__, ret, pkt);
+	} else {
+		struct nrf91_non_offload_dev_context *ctx = dev->data;
+
+		ret = nrf_send(ctx->mdm_skct_id, send_buffer, data_len, 0);
+		if (ret < 0) {
+			mosh_error("%s: send() failed: (%d), data len: %d\n", __func__, -errno,
+			       data_len);
+		} else if (ret != data_len) {
+			mosh_error("%s: only partially sent, only %d of original %d was sent",
+				__func__,
+				ret, data_len);
+		}
+	}
+	net_pkt_unref(pkt);
+
+	return ret;
+}
+
+/******************************************************************************/
+
+#define NRF91_MODEM_DATA_DL_POLL_TIMEOUT_MS 1000
+#define NRF91_MODEM_DATA_DL_BUFFER_SIZE 1500
+
+static char receive_buffer[NRF91_MODEM_DATA_DL_BUFFER_SIZE];
+
+static void nrf91_non_offload_modem_dl_data_thread_handler(void)
+{
+	struct nrf_pollfd fds[1];
+	struct net_pkt *rcv_pkt;
+
+	int ret = 0;
+	int recv_data_len = 0;
+
+	while (true) {
+		if (nrf91_non_offload_iface_data.mdm_skct_id < 0) {
+			/* Wait for sockets to be created */
+			k_sem_take(&mdm_socket_sem, K_FOREVER);
+			continue;
+		}
+		fds[0].fd = nrf91_non_offload_iface_data.mdm_skct_id;
+		fds[0].events = NRF_POLLIN;
+		fds[0].revents = 0;
+
+		ret = nrf_poll(fds, 1, NRF91_MODEM_DATA_DL_POLL_TIMEOUT_MS);
```

Comment:
What do we need `poll()` and timeout for here? Can't we just block in `nrf_recv()` waiting for data to come?

Reply:
No need. poll() was now removed.